core/commonMain/src/implementations/immutableMap/TrieNode.kt

```diff
@@ -477,6 +562,103 @@ internal class TrieNode<K, V>(
+                TrieNode(newDataMap, newNodeMap, newBuffer)
+            }
+        }
+        newNodeMap.forEachOneBit { positionMask, index ->
+            val newNodeIndex = mutableNode.buffer.size - 1 - index
+            mutableNode.buffer[newNodeIndex] = when {
+                hasNodeAt(positionMask) -> {
+                    val before = nodeAtIndex(nodeIndex(positionMask))
+                    before.mutablePutAllFromOtherNodeCell(otherNode, positionMask, shift, intersectionCounter, mutator)
+                }
+
+                otherNode.hasNodeAt(positionMask) -> {
+                    val before = otherNode.nodeAtIndex(otherNode.nodeIndex(positionMask))
+                    before.mutablePutAllFromOtherNodeCell(this, positionMask, shift, intersectionCounter, mutator)
+                }
+
+                else -> { // two entries, and they are not equal by key (see ** above)
+                    val thisKeyIndex = this.entryKeyIndex(positionMask)
+                    val thisKey = this.keyAtIndex(thisKeyIndex)
+                    val thisValue = this.valueAtKeyIndex(thisKeyIndex)
+                    val otherKeyIndex = otherNode.entryKeyIndex(positionMask)
+                    val otherKey = otherNode.keyAtIndex(otherKeyIndex)
+                    val otherValue = otherNode.valueAtKeyIndex(otherKeyIndex)
+                    makeNode(
+                            thisKey.hashCode(),
+                            thisKey,
+                            thisValue,
+                            otherKey.hashCode(),
+                            otherKey,
+                            otherValue,
+                            shift + LOG_MAX_BRANCHING_FACTOR,
+                            mutator.ownership
+                    )
+                }
+            }
+        }
+        newDataMap.forEachOneBit { positionMask, index ->
+            val newKeyIndex = index * ENTRY_SIZE
+            when {
+                !otherNode.hasEntryAt(positionMask) -> {
+                    val oldKeyIndex = this.entryKeyIndex(positionMask)
+                    mutableNode.buffer[newKeyIndex] = this.keyAtIndex(oldKeyIndex)
+                    mutableNode.buffer[newKeyIndex + 1] = this.valueAtKeyIndex(oldKeyIndex)
+                }
+                // there is either only one entry in otherNode, or
+                // both entries are here => they are equal, see ** above
+                // so just overwrite that
+                else -> {
+                    val oldKeyIndex = otherNode.entryKeyIndex(positionMask)
+                    mutableNode.buffer[newKeyIndex] = otherNode.keyAtIndex(oldKeyIndex)
+                    mutableNode.buffer[newKeyIndex + 1] = otherNode.valueAtKeyIndex(oldKeyIndex)
+                    if (this.hasEntryAt(positionMask)) intersectionCounter.count++
+                }
+            }
+        }
+        return when {
+            this.elementsEquals(mutableNode) -> this
```

Comment:
Let's rename the method to `elementsIdentityEquals` to have consistent naming with set `TrieNode`